radixdlt-core/radixdlt/src/main/java/com/radixdlt/store/berkeley/BerkeleyLedgerEntryStore.java

```diff
@@ -335,165 +400,178 @@ public void commit(AID aid) {
+	}
+
+	private void removeLastPending(com.sleepycat.je.Transaction transaction) {
+		try (var cursor = pendingDatabase.openCursor(transaction, null)) {
+			var pKey = entry();
+			var value = entry();
+			var status = cursor.getLast(pKey, value, DEFAULT);
+
+			if (status == SUCCESS) {
+				addBytesRead(value, pKey);
+				pendingDatabase.delete(transaction, pKey);
 			}
 		} catch (Exception e) {
 			transaction.abort();
 			fail("Commit of atom failed", e);
 		}
-
-		DatabaseEntry vertexKey = new DatabaseEntry(vertexStoreState.getRoot().getId().asBytes());
-		SerializedVertexStoreState serializedVertexWithQC = vertexStoreState.toSerialized();
-		DatabaseEntry vertexEntry = new DatabaseEntry(serialization.toDson(serializedVertexWithQC, Output.ALL));
-		OperationStatus putStatus = this.pendingDatabase.put(transaction, vertexKey, vertexEntry);
-		if (putStatus != OperationStatus.SUCCESS) {
-			fail("Store of root vertex failed");
-		} else {
-			addBytesWrite(vertexEntry.getSize() + vertexKey.getSize());
-		}
 	}
 
 	private LedgerEntryStoreResult doStore(
 		long logicalClock,
 		AID aid,
-		byte[] ledgerEntryData,
+		DatabaseEntry pData,
+		boolean fullData,
 		LedgerEntryIndices indices,
 		com.sleepycat.je.Transaction transaction
 	) throws DeserializeException {
+		var offset = 0L;
 		try {
-			DatabaseEntry pKey = toPKey(BerkeleyLedgerEntryStore.PREFIX_COMMITTED, logicalClock, aid);
-			DatabaseEntry pData = new DatabaseEntry(ledgerEntryData);
-
 			// put indices in temporary map for key creator to pick up
-			this.currentIndices.put(aid, indices);
-			OperationStatus status = this.atoms.putNoOverwrite(transaction, pKey, pData);
-			if (status != OperationStatus.SUCCESS) {
-				fail("Atom write for '" + aid + "' failed with status " + status);
+			currentIndices.put(aid, indices);
+
+			var pKey = toPKey(PREFIX_COMMITTED, logicalClock, aid);
+
+			if (fullData) {
+				var logData = entry(new byte[Long.BYTES]);
+				offset = atomLog.position();
+
+				Longs.copyTo(offset, logData.getData(), 0);
+				failIfNotSuccess(atoms.putNoOverwrite(transaction, pKey, logData), "Atom write for", aid);
 			} else {
-				addBytesWrite(ledgerEntryData.length + pKey.getSize());
+				failIfNotSuccess(atoms.putNoOverwrite(transaction, pKey, pData), "Atom write for", aid);
 			}
+			addBytesWrite(pData, pKey);
 
-			DatabaseEntry indicesData = new DatabaseEntry(serialization.toDson(indices, Output.PERSIST));
-			status = this.atomIndices.putNoOverwrite(transaction, pKey, indicesData);
-			if (status != OperationStatus.SUCCESS) {
-				fail("LedgerEntry indices write for '" + aid + "' failed with status " + status);
-			} else {
-				addBytesWrite(indicesData.getSize() + pKey.getSize());
+			var indicesData = serialize(indices);
+
+			failIfNotSuccess(atomIndices.putNoOverwrite(transaction, pKey, indicesData), "LedgerEntry indices write for", aid);
+			addBytesWrite(indicesData, pKey);
+
+			if (fullData) {
+				atomLog.write(pData.getData());
```

Comment:
There is no such mechanism yet. As well as there is no index restoring mechanism in case if we're starting with existing log. Those mechanisms require separate efforts. I've created new task for this: https://radixdlt.atlassian.net/browse/RPNV1-1020